### PR TITLE
Eve's side

### LINKED ISSUE #16: Show quiz to the Eve
-  Build page 11(Eve view of quiz asking for first_name, last_name and email)
-  Build page 12(Eve view of quiz for submitting quiz answers)
-  Build page 13(quiz result page showing Eve's answer along with the correct answer for each question)
-  For the Eve, we can create user(standard) record as Eve Smith in the existing user table
- Eve can attempt for a quiz only once.
-  Backend Design
-  Create a table `attempts` for storing `attempt record` for a Eve which has a quiz as reference.
-  Add one boolean column `submitted` in `attempts` table as well.
-  Create another table `attempt_answers` for storing answers. Each record of the table contains answer for single question of an attempt.
-   In table `attempt_answers`, combination of `attempt_id` and `question_id` will be unique.

**Workflow for Eve**

1. Eve enters a valid quiz slug.
2. Slug is verified and redirects to `attempts/new` page.
3. Eve enters its details and submits it to get the quiz details.
4. Eve is created in the user's table as a `standard` user and an `attempt` record is created with submitted as false.
5. Eve answers the questions and `submitted` is set `true` in the `attempt` table with all the answers in `attempt_answers` table. (End of Eve's flow)
6. On step 3-4 if Eve refreshes the page Eve will be asked to enter the user details again.
7. If we find the details of Eve with an `unsubmitted(submitted: false)` attempt then reuse the user object or else create a new user.
8. Reports should contain only submitted attempts.

### PR DESCRIPTION
**Pull request by josephv7**
Fixes #16

### CODE REVIEW

File: app/controllers/api/v1/attempts_controller.rb [hunk 1]
```diff
@@ -0,0 +1,44 @@
+class Api::V1::AttemptsController < Api::V1::BaseController
+  skip_before_action :authenticate_user_using_x_auth_token, only: [:create, :update]
+  before_action :load_attempt, only: [:update]
+  def create
+
+    @attempt = Attempt.where(attempt_params).first
```
Comment (rafi-gemhar33) [inline app/controllers/api/v1/attempts_controller.rb:6]: FYI `Attempt.where(attempt_params).first` => `Attempt.find_by(attempt_params)`

File: app/controllers/api/v1/attempts_controller.rb [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    if @attempt.nil?
+      @attempt_user = Attempt.new(attempt_params.merge(submitted: "false"))
+      p @attempt_user
```
Comment (rafi-gemhar33) [inline app/controllers/api/v1/attempts_controller.rb:9]: remove logs

File: app/controllers/api/v1/attempts_controller.rb [hunk 3]
```diff
@@ -0,0 +1,44 @@
+class Api::V1::AttemptsController < Api::V1::BaseController
+  skip_before_action :authenticate_user_using_x_auth_token, only: [:create, :update]
+  before_action :load_attempt, only: [:update]
+  def create
+
+    @attempt = Attempt.where(attempt_params).first
+    if @attempt.nil?
+      @attempt_user = Attempt.new(attempt_params.merge(submitted: "false"))
```
Comment (rafi-gemhar33) [inline app/controllers/api/v1/attempts_controller.rb:8]: why is `Attempt.new `saved in `@attempt_user`.

All this logic could be moved to an Attempt::CreateService.
With the pseudo code something like:
  ```
  existing_attempt = Attempt.find_by(create_attempt_params)
    if existing_attempt.present? 
        if existing_attempt.submitted
          # { errors: ["Quiz attempt already submitted"] }
        else
          # { notice: "Quiz attempt already exists", attempt: existing_attempt }
        end
        return
    end

    new_attempt = @user.attempts.new(create_attempt_params)
    if new_attempt.save
     # { notice: "Attempt created", attempt: new_attempt }
    else
      # { errors: new_attempt.errors.full_messages }
    end
```

File: app/controllers/api/v1/attempts_controller.rb [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+        render status: :unprocessable_entity, json:{ errors: @attempt_user.errors.full_messages }
+      end
+    else
+      render status: :ok, json: { notice: "Attempt Already Exists" }
+    end
+
+  end
+
+  def update
+    if @attempt.update(attempt_update_params)
```
Comment (rafi-gemhar33) [inline app/controllers/api/v1/attempts_controller.rb:22]: have a check that if @attempt is submitted then do not update again.
    if @attempt.submitted
      render status: :forbidden, json: { notice: "Quiz attempt already submitted" }
      return    end

File: app/controllers/api/v1/attempts_controller.rb [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+      render status: :ok, json: { notice: "Record Updated After Submission." }
+    else
+      render status: :unprocessable_entity, json:{ errors: @attempt.errors.full_messages }
+    end
+  end
+
+  private
+    def attempt_params
```
Comment (rafi-gemhar33) [inline app/controllers/api/v1/attempts_controller.rb:30]: rename it to `attempt_create_params` for clarity

File: app/controllers/api/v1/quiz_controller.rb [hunk 6]
```diff
@@ -30,6 +30,13 @@ def publish
 
   end
 
+  def check_slug
```
Comment (rafi-gemhar33) [inline app/controllers/api/v1/quiz_controller.rb:33]: avoid custom actions in controller

File: app/controllers/api/v1/users_controller.rb [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+        render status: :unprocessable_entity, json:{ errors: @user.errors.full_messages }
+      else
+        render json: { notice: "User Created", authenticate: true, userId: @user['id'] }, status: :ok
+      end
+
+  end
+
+  private
+    def user_params
+      params.require(:user).permit(:email, :first_name, :last_name, :slug)
```
Comment (rafi-gemhar33) [inline app/controllers/api/v1/users_controller.rb:31]: please change the params structure.

```
 {
 user: {email, :first_name, :last_name},
 quiz: {slug}
 }
```

**Comment by rafi-gemhar33** [comment]
@josephv7 
Please add unique validation for slug in model & db, I mised this

File: test/models/attempt_answer_test.rb [hunk 8]
```diff
@@ -0,0 +1,7 @@
+require 'test_helper'
```
[the remark this answers is not in the document]
Comment (josephv7) [reply test/models/attempt_answer_test.rb:1]: Will fix in a separate issue.

**Pull request merged**